File: daily/Sliding Window/1438-Longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.go
package daily

func longestSubarray(nums []int, limit int) int {
	maxDeque := []int{}
	minDeque := []int{}

	n := len(nums)
	left := 0
	res := 0

	for right := 0; right < n; right++ {
		// Maintain maxDeque (decreasing)
		for len(maxDeque) > 0 && nums[maxDeque[len(maxDeque)-1]] < nums[right] {
			maxDeque = maxDeque[:len(maxDeque)-1]
		}
		maxDeque = append(maxDeque, right)

		// Maintain minDeque (increasing)
		for len(minDeque) > 0 && nums[minDeque[len(minDeque)-1]] > nums[right] {
			minDeque = minDeque[:len(minDeque)-1]
		}
		minDeque = append(minDeque, right)

		// Shrink window if condition violated
		for nums[maxDeque[0]]-nums[minDeque[0]] > limit {
			left++
			if maxDeque[0] < left {
				maxDeque = maxDeque[1:]
			}
			if minDeque[0] < left {
				minDeque = minDeque[1:]
			}
		}

		if right-left+1 > res {
			res = right - left + 1
		}
	}
	return res
}

File: daily/Sliding Window/1438-Longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.py
class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        max_deque = deque()
        min_deque = deque()
        
        n = len(nums)
        left = res = 0
        for right in range(n):
            while max_deque and nums[max_deque[-1]] < nums[right]:
                max_deque.pop()
            max_deque.append(right)

            while min_deque and nums[min_deque[-1]] > nums[right]:
                min_deque.pop()
            min_deque.append(right)  

            while nums[max_deque[0]] - nums[min_deque[0]] > limit:
                left += 1
                if max_deque[0] < left:
                    max_deque.popleft()
                if min_deque[0] < left:
                    min_deque.popleft() 

            res = max(res, right - left + 1)
        return res
